pkg/gatewayserver/io/ws/ws.go

```diff
@@ -321,23 +335,24 @@ func (s *srv) handleTraffic(w http.ResponseWriter, r *http.Request) (err error)
 			case <-conn.Context().Done():
 				return
 			case <-pingTicker.C:
-				if err := ws.WriteMessage(websocket.PingMessage, nil); err != nil {
```

Comment:
In the old code, notice that between the moment in which we finish sending the ping, and the moment in which we increment the number of missing pongs, anything can happen, including receiving a pong. This is the crux of the flake - if you have zero missed pongs, you receive a pong so you replace zero with zero, you will accidentally create one missing pong.

The error flake is created by the following sequence of events.

1. Send a ping to the client: `missingPongs == -1`.
3. Receive a pong from the client: `missingPongs == 0`.
4. Increment the number of missing pongs: `missingPogs == 1`.
5. Send a ping to the client: `missingPongs == 1`.
6. Increment the number of missing pongs: `missingPongs == 2`.
7. You are now disconnecting the gateway while in reality only one pong is missing.

The proposed solution is to increment the number of missing pongs before sending the pings, and using the `missingPongs` counter as the difference between the outstanding pings and pongs. The counter is not allowed to go negative, as it implies that we have received more pongs than pings, which is a signal of client implementation error.